src/cursor/mod.rs

```diff
@@ -21,101 +30,73 @@ use crate::{error::Result, operation::GetMore, options::StreamAddress, Client, N
+/// Additionally, all the other methods that an [`Stream`](https://docs.rs/futures/0.3.4/futures/stream/trait.Stream.html) has are available on `Cursor` as well.
+/// This includes all of the functionality provided by [`StreamExt`](https://docs.rs/futures/0.3.4/futures/stream/trait.StreamExt.html), which provides similar functionality to the standard library `Iterator` trait.
 /// For instance, if the number of results from a query is known to be small, it might make sense
 /// to collect them into a vector:
 ///
 /// ```rust
 /// # use bson::{doc, bson, Document};
+/// # use futures::stream::StreamExt;
 /// # use mongodb::{Client, error::Result};
 /// #
 /// # async fn do_stuff() -> Result<()> {
 /// # let client = Client::with_uri_str("mongodb://example.com").await?;
 /// # let coll = client.database("foo").collection("bar");
 /// # let cursor = coll.find(Some(doc! { "x": 1 }), None)?;
 /// #
-/// let results: Vec<Result<Document>> = cursor.collect();
+/// let results: Vec<Result<Document>> = cursor.collect().await;
 /// # Ok(())
 /// # }
 /// ```
 #[derive(Debug)]
 pub struct Cursor {
-    client: Client,
-    buffer: VecDeque<Document>,
-    get_more: GetMore,
-    exhausted: bool,
+    inner: ImpatientCursor,
+    tailable: bool,
 }
 
 impl Cursor {
-    pub(crate) fn new(client: Client, spec: CursorSpecification) -> Self {
-        let get_more = GetMore::new(
-            spec.ns,
-            spec.id,
-            spec.address,
-            spec.batch_size,
-            spec.max_time,
-        );
-
+    pub(crate) fn new(client: Client, spec: CursorSpecification, tailable: bool) -> Self {
         Self {
-            client,
-            buffer: spec.buffer,
-            get_more,
-            exhausted: spec.id == 0,
-        }
-    }
-}
-
-impl Drop for Cursor {
-    fn drop(&mut self) {
-        if self.exhausted {
-            return;
+            inner: ImpatientCursor::new(client, spec),
+            tailable,
         }
-
-        let namespace = self.get_more.namespace();
-
-        let _ = RUNTIME.block_on(self.client.database(&namespace.db).run_command(
-            doc! {
-                "killCursors": &namespace.coll,
-                "cursors": [self.get_more.cursor_id()]
-            },
-            None,
-        ));
     }
 }
 
-impl Iterator for Cursor {
+impl Stream for Cursor {
     type Item = Result<Document>;
 
-    fn next(&mut self) -> Option<Self::Item> {
-        if self.buffer.is_empty() && !self.exhausted {
-            match RUNTIME.block_on(self.client.execute_operation(&self.get_more, None)) {
-                Ok(get_more_result) => {
-                    self.buffer.extend(get_more_result.batch);
-                    self.exhausted = get_more_result.exhausted;
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        if self.tailable {
```

Comment:
I don't think we need a special flag for tailable. I think we can instead just check if the underlying cursor is still alive. As it is right now, once a tailable cursor dies / is exhausted, this will loop forever.

As a side note, we should probably return an error if a user tries to iterate a dead cursor. We have to be careful not to return an error if the cursor dies after we start repeatedly polling it though.

Reply:
I've made the suggested changes except for the part about returning an error when the cursor is exhausted. Is this something we really want to do? I don't think I've seen any other iterators in Rust do that.

Reply:
I just mentioned it since it's something we do in Swift. Note that I'm only referring to returning errors when beginning a new iteration the cursor after it has indicated it is done, not just whenever it becomes exhausted a la Python.The rationale for doing so is that a user iterating a dead cursor is probably indicative of a programmer error:
- In the "patient" case, it could only happen if they're not using a `while let` loop or if they start iterating after their `while let` loop ended, both likely errors.
- In the "non-patient" case, it means they're not distinguishing between a cursor being exhausted and returning `None` (like we were doing here), which is also usually an error since it can lead to infinite loops.

All that being said, I'm completely fine with not including it here and discussing it later.

Reply:
I've filed [RUST-337](https://jira.mongodb.org/browse/RUST-337) to discuss it.